src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -52,63 +52,88 @@ public async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, Can
+                        using (Http2WriteStream writeStream = new Http2WriteStream(this))
+                        {
+                            await request.Content.CopyToAsync(writeStream).ConfigureAwait(false);
+                        }
                     }
-                }
 
-                // Wait for response headers to be read.
-                await readDataAvailableTask.ConfigureAwait(false);
+                    // Wait for response headers to be read.
+                    await readDataAvailableTask.ConfigureAwait(false);
 
-                // Start to process the response body.
-                bool emptyResponse = false;
-                lock (_syncObject)
-                {
-                    if (_responseComplete && _responseBuffer.ActiveSpan.Length == 0)
+                    // Start to process the response body.
+                    bool emptyResponse = false;
+                    lock (_syncObject)
                     {
-                        if (_responseAborted)
+                        if (_responseComplete && _responseBuffer.ActiveSpan.Length == 0)
                         {
-                            throw new IOException(SR.net_http_invalid_response);
+                            if (_responseAborted)
+                            {
+                                throw new IOException(SR.net_http_invalid_response);
+                            }
+
+                            emptyResponse = true;
                         }
+                    }
 
-                        emptyResponse = true;
+                    if (emptyResponse)
+                    {
+                        responseContent.SetStream(EmptyReadStream.Instance);
+                    }
+                    else
+                    {
+                        responseContent.SetStream(new Http2ReadStream(this));
                     }
-                }
 
-                if (emptyResponse)
-                {
-                    responseContent.SetStream(EmptyReadStream.Instance);
+                    // Process Set-Cookie headers.
+                    if (_connection._pool.Settings._useCookies)
+                    {
+                        CookieHelper.ProcessReceivedCookies(_response, _connection._pool.Settings._cookieContainer);
+                    }
                 }
-                else
+                catch (Exception e)
                 {
-                    responseContent.SetStream(new Http2ReadStream(this));
-                }
+                    Dispose();
 
-                // Process Set-Cookie headers.
-                if (_connection._pool.Settings._useCookies)
-                {
-                    CookieHelper.ProcessReceivedCookies(_response, _connection._pool.Settings._cookieContainer);
+                    if (e is IOException ioe)
+                    {
+                        throw new HttpRequestException(SR.net_http_client_execution_error, ioe);
```

Comment:
Curious why we surface the original exception as an inner exception for IOException but not for the others?

Reply:
We started embedded HTTP related exceptions inside IOException in .NET Framework when we wrote HttpClientHandler.  The System.IO.Stream returned by HTTP APIs needs to return IOException as the top-level exception.  But we wanted to provide more information about the root cause.  That is how this practice started.

I do think it would be useful to embed these root cause exceptions in other places.

Which places do you think need that?